src/librustc_mir/borrow_check/nll/type_check/relate_tys.rs

```diff
@@ -262,76 +261,52 @@ impl<'cx, 'bccx, 'gcx, 'tcx> TypeRelating<'cx, 'bccx, 'gcx, 'tcx> {
-            // been "instantiated" and hence the set of scopes we have
-            // doesn't matter -- just to be sure, put an empty vector
-            // in there.
-            let old_a_scopes = mem::replace(&mut self.a_scopes, vec![]);
-            let result = self.relate(&a_kind, &b_kind);
-            self.a_scopes = old_a_scopes;
+        // The generalized values we extract from `canonical_var_values` have
+        // been fully instantiated and hence the set of scopes we have
+        // doesn't matter -- just to be sure, put an empty vector
+        // in there.
+        let old_a_scopes = ::std::mem::replace(&mut self.a_scopes, vec![]);
 
-            debug!("equate_var: complete, result = {:?}", result);
-            return result;
-        }
+        // Relate the generalized kind to the original one.
+        let result = self.relate(&generalized_kind, &b_kind);
 
-        // Not yet. Capture the value from the RHS and carry on.
-        let closed_kind =
-            self.instantiate_traversed_binders(universal_regions, &self.b_scopes, b_kind);
-        self.canonical_var_values[var] = Some(closed_kind);
-        debug!(
-            "equate_var: capturing value {:?}",
-            self.canonical_var_values[var]
-        );
+        // Restore the old scopes now.
+        self.a_scopes = old_a_scopes;
 
-        // FIXME -- technically, we should add some sort of
-        // assertion that this value can be named in the universe
-        // of the canonical variable. But in practice these
-        // canonical variables only arise presently in cases where
-        // they are in the root universe and the main typeck has
-        // ensured there are no universe errors. So we just kind
-        // of over look this right now.
-        Ok(b_kind)
+        debug!("equate_var: complete, result = {:?}", result);
+        return result;
     }
 
-    /// As we traverse types and pass through binders, we push the
-    /// values for each of the regions bound by those binders onto
-    /// `scopes`. This function goes through `kind` and replaces any
-    /// references into those scopes with the corresponding free
-    /// region. Thus the resulting value should have no escaping
-    /// references to bound things and can be transported into other
-    /// scopes.
-    fn instantiate_traversed_binders(
+    fn generalize_value(
         &self,
-        universal_regions: &UniversalRegions<'tcx>,
-        scopes: &[BoundRegionScope],
         kind: Kind<'tcx>,
     ) -> Kind<'tcx> {
-        let k = kind.fold_with(&mut BoundReplacer {
+        TypeGeneralizer {
             type_rel: self,
             first_free_index: ty::INNERMOST,
-            universal_regions,
-            scopes: scopes,
-        });
+            ambient_variance: self.ambient_variance,
 
-        assert!(!k.has_escaping_regions());
-
-        k
+            // These always correspond to an `_` or `'_` written by
```

Comment:
I just want to make sure I'm understanding this comment: You are drawing this conclusion that the type (or region) is always either `_` or `'_` ... because that is the only scenario where `canonical_var_values[var]` is allowed to return `None`?